File: 872.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	l1 := make([]int, 0)
	l2 := make([]int, 0)
	dfs(&l1, root1)
	dfs(&l2, root2)

	if len(l1) != len(l2) {
		return false
	}
	for i := 0; i < len(l1); i++ {
		if l1[i] != l2[i] {
			return false
		}
	}
	return true
}

func dfs(l *[]int, root *TreeNode) {
	if root == nil {
		return
	}
	if root.Left == nil && root.Right == nil {
		*l = append(*l, root.Val)
		return
	}
	dfs(l, root.Left)
	dfs(l, root.Right)
}

File: 872.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> l1 = new ArrayList<Integer>();
        List<Integer> l2 = new ArrayList<Integer>();

        dfs(l1, root1);
        dfs(l2, root2);

        return l1.equals(l2);
    }

    public void dfs(List<Integer> l, TreeNode root) {
        if(root == null) return ;
        if(root.left == null && root.right == null) {
            l.add(root.val);
        }
        dfs(l, root.left);
        dfs(l, root.right);
    }
}